core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -0,0 +1,267 @@
+            this.request = request;
+            /*
+             * We call this before replication because this might wait for a refresh and that can take a while. This way we wait for the
+             * refresh in parallel on the primary and on the replica.
+             */
+            finishWrite(indexShard, request, location);
+        }
+
+        @Override
+        public synchronized void respond(ActionListener<Response> listener) {
+            this.listener = listener;
+            respondIfPossible();
+        }
+
+        /**
+         * Respond if the refresh has occurred and the listener is ready. Always called while synchronized on {@code this}.
+         */
+        protected void respondIfPossible() {
+            if (finishedWrite && listener != null) {
+                super.respond(listener);
+            }
+        }
+
+        @Override
+        public void forcedRefresh() {
+            finalResponse.setForcedRefresh(true);
+        }
+
+        @Override
+        public void waitForRefreshForcedRefresh() {
+            logger.warn("block_until_refresh request ran out of slots and forced a refresh: [{}]", request);
+        }
+
+        @Override
+        public synchronized void finished() {
+            finishedWrite = true;
+            respondIfPossible();
+        }
+    }
+
+    /**
+     * Result of taking the action on the replica.
+     */
+    private class WriteReplicaResult extends ReplicaResult implements RespondingWriteResult<TransportResponse.Empty> {
+        private final IndexShard indexShard;
+        private final ReplicatedWriteRequest<?> request;
+        private final Translog.Location location;
+        private ActionListener<TransportResponse.Empty> listener;
+
+        public WriteReplicaResult(IndexShard indexShard, ReplicatedWriteRequest<?> request, Translog.Location location) {
+            this.indexShard = indexShard;
+            this.request = request;
+            this.location = location;
+        }
+
+        @Override
+        public void respond(ActionListener<TransportResponse.Empty> listener) {
+            this.listener = listener;
+            finishWrite(indexShard, request, location);
+        }
+
+        @Override
+        public void forcedRefresh() {
+            // It'd be nice to mark this in the response but logging when wait for refresh forced it is OK.
+        }
+
+        @Override
+        public void waitForRefreshForcedRefresh() {
+            logger.warn("block_until_refresh request ran out of slots and forced a refresh: [{}]", request);
+        }
+
+        @Override
+        public void finished() {
+            listener.onResponse(TransportResponse.Empty.INSTANCE);
+        }
+    }
+
+    /**
+     * Duplicate code shared between WritePrimaryResult and WriteReplicaResult. Implemented as an interface because it lets us instantiate a
+     * few fewer classes during this process. Package private for testing.
+     */
+    interface RespondingWriteResult<R> {
```

Comment:
I see how you solved the multiple inheritance issue we want here with an interface and default implementation. Neat. I do however think it can be simpler with a static method. I prepared a patch to show what I mean - "simpler" is subjective :)